src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs

```diff
@@ -181,80 +194,77 @@ public async Task<T> OpenContextForWriteAsync<T>(Func<IWorkspaceProjectContextAc
                 }
             }
 
-            internal Task OnProjectChangedAsync(ProjectChange change, WorkspaceContextHandlerType handlerType)
+            internal Task OnProjectChangedAsync(
+                IDataProgressTrackerServiceRegistration registration,
+                ConfiguredProject activeConfiguredProject,
+                StrongBox<ContextState?> lastContextState,
+                IImmutableDictionary<NamedIdentity, IComparable> dataSourceVersions,
+                Func<bool> hasChange,
+                Action<ContextState, CancellationToken> applyFunc)
             {
-                return ExecuteUnderLockAsync(ct => ApplyProjectChangesUnderLockAsync(change, handlerType, ct), _tasksService.UnloadCancellationToken);
-            }
+                return ExecuteUnderLockAsync(ApplyProjectChangesUnderLockAsync, _tasksService.UnloadCancellationToken);
 
-            private async Task ApplyProjectChangesUnderLockAsync(ProjectChange change, WorkspaceContextHandlerType handlerType, CancellationToken cancellationToken)
-            {
-                // NOTE we cannot call CheckForInitialized here, as this method may be invoked during initialization
-                Assumes.NotNull(_contextAccessor);
+                Task ApplyProjectChangesUnderLockAsync(CancellationToken cancellationToken)
+                {
+                    // NOTE we cannot call CheckForInitialized here, as this method may be invoked during initialization
+                    Assumes.NotNull(_contextAccessor);
 
-                bool isActiveEditorContext = _activeWorkspaceProjectContextTracker.IsActiveEditorContext(_contextAccessor.ContextId);
-                bool isActiveConfiguration = change.ActiveConfiguredProject == _project;
+                    (ContextState contextState, bool stateChanged) = GetContextState(lastContextState);
 
-                bool hasChange =
-                    change.Subscription.Value.ProjectChanges.Any(c => c.Value.Difference.AnyChanges) ||
-                    change.CommandLineArgumentsSnapshot?.IsChanged == true ||
-                    _lastContextState?.IsActiveConfiguration != isActiveConfiguration ||
-                    _lastContextState?.IsActiveEditorContext != isActiveEditorContext;
+                    if (!stateChanged && !hasChange())
```

Comment:
I'm unconvinced we need to handle the context change, in fact, I think you'll hit the file name caches within the handlers themselves which do not persist this state.

Reply:
This is also not tracking it across evaluation/build results, so the logic is wrong before and after your change as one will reset the other.

Reply:
Ah, I'm wrong for the last point, but I would verify that we're even doing anything as per the first point.

Reply:
Regarding your last point, that's how it used to behave -- with one `lastContext` value for all three updates, which seemed like a bug for any handler that cared about seeing these values change state.

Will investigate further on your first point.

Reply:
Ok I see what you mean, the `isActiveContext` value is only used when the item is first added to the workspace in the current implementation.

Is that a bug that needs fixing, or should we just remove the change tracking here?

Reply:
Remove it, I'm pretty sure Roslyn handles this correctly on the switch. This is just whether Roslyn goes looking for open documents the first time the file appears.

Reply:
https://github.com/dotnet/project-system/pull/7862